src/request.rs

```diff
@@ -277,6 +277,47 @@ impl RequestBuilder {
         Ok(self)
     }
 
+    /// Sends a multipart/formdata body.
+    ///
+    /// ```no_run
+    /// use reqwest::mime;
+    /// use reqwest::{MultipartRequest, MultipartField, to_multipart};
+    ///
+    /// let client = reqwest::Client::new().unwrap();
+    /// let response = client.post("your url").unwrap()
+    ///     .multipart(
+    ///         // Add fields from anything serializable
+    ///         to_multipart(&[("key", "value"), ("key2", "value2")]).unwrap()
+    ///         // Add fields builder style
+    ///         .field(MultipartField::param("key3", "value3"))
+    ///         .field(MultipartField::param("json", "{ \"number\": 5 }")
+    ///             .mime(Some(mime::APPLICATION_JSON)))
+    ///         .field(MultipartField::file("file", "/path/to/file")
+    ///             .expect("File could not be opened"))
+    ///     ).send();
+    ///
+    /// ```
+    ///
+    /// See [`to_multipart`](fn.to_multipart.html), [`MultipartRequest`](struct.MultipartRequest.html)
+    /// and [`MultipartField`](struct.MultipartField.html) for more examples.
+    pub fn multipart(&mut self, multipart: MultipartRequest) -> &mut RequestBuilder {
```

Comment:
> I gave this some more thought. How are nested values serialized? Does it serialize to a thing of name = "products[bike][name]" or something? I'm not sure if that is usual, or just something that some individuals do.

I dont quite understand this part.

> Anywho, if nested fields don't work, then we could instead make the serialize look certain structure patterns, kind of like requests does. Either with use of a specific order in tuples, or by blessing certain names... Examples:

What you describe seems possible but are you sure it is not too convoluted for what it brings? Not only would it make the serializer a lot more complex (at least I think it would but again this was my first time implementing a Serde trait) but it also makes it unclear what kind of serializeable things are valid and lead to what kind of multipart values since the description of that only exists in the serializer implementation and then has to be communicated to the user over the documentation. It is not type checked and might fail or do unexpected things when run.

The current serializer makes it easy to add simple parameters. For more complex things it might even be harder to use them with serialize because the builder methods are very clear in what they do whereas what exactly the serializer might do is harder to understand.